jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java

```diff
@@ -1779,6 +1757,119 @@ public void testTriggerNotExistingAdHocFragment() {
     }
 
     @Test
+    public void testTriggerAdHocTasksFromCurrentAndNextStage() {
+        identityProvider.setName("john");
+        Map<String, OrganizationalEntity> roleAssignments = new HashMap<>();
+        roleAssignments.put("owner", new UserImpl("john"));
+
+        Map<String, Object> data = new HashMap<>();
+        data.put("customData", "none");
+
+        CaseFileInstance caseFile = caseService.newCaseFileInstance(deploymentUnit.getIdentifier(), TWO_STAGES_CONDITIONS_CASE_P_ID, data, roleAssignments);
+        String caseId = caseService.startCase(deploymentUnit.getIdentifier(), TWO_STAGES_CONDITIONS_CASE_P_ID, caseFile);
+        assertNotNull(caseId);
+
+        try {
+            Collection<CaseStageInstance> stages = caseRuntimeDataService.getCaseInstanceStages(caseId, false, null);
+            assertThat(stages).isNotNull().hasSize(2);
+            Iterator<CaseStageInstance> iterator = stages.iterator();
+
+            CaseStageInstance stage1 = iterator.next();
+            assertThat(stage1.getName()).isEqualTo("Stage One");
+            assertThat(stage1.getStatus()).isEqualTo(StageStatus.Active);
+
+            CaseStageInstance stage2 = iterator.next();
+            assertThat(stage2.getName()).isEqualTo("Stage Two");
+            assertThat(stage2.getStatus()).isEqualTo(StageStatus.Available);
+
+            List<TaskSummary> tasks = runtimeDataService.getTasksAssignedAsPotentialOwner("john", new QueryFilter());
+            assertThat(tasks).isNotNull().isEmpty();
+
+            caseService.triggerAdHocFragment(caseId, "Task 1", null);
+            tasks = runtimeDataService.getTasksAssignedAsPotentialOwner("john", new QueryFilter());
+            assertThat(tasks).isNotNull().hasSize(1);
+            assertTask(tasks.get(0), "john", "Task 1", Status.Reserved);
+
+            caseService.triggerAdHocFragment(caseId, "Task 2", null);
+            // the task from the next stage should not be triggered
```

Comment:
@mswiderski I am not sure if the current behavior here is correct. Should not we throw `AdHocFragmentNotFoundException` if you try to trigger an ad-hoc fragment from the stage which is not active?